utilities/rate_limiters/write_amp_based_rate_limiter.cc

```diff
@@ -331,21 +338,33 @@ Status WriteAmpBasedRateLimiter::Tune() {
 
   int64_t prev_bytes_per_sec = GetBytesPerSecond();
 
-  // Loop through the actual time slice to make sure bytes flow from long period
-  // of time is properly estimated when the compaction rate is low.
+  // This function can be called less frequent than we anticipate when
+  // compaction rate is low. Loop through the actual time slice to correct
+  // the estimation.
   for (uint32_t i = 0; i < duration_ms / kMillisPerTune; i++) {
     bytes_sampler_.AddSample(duration_bytes_through_ * 1000 / duration_ms);
     highpri_bytes_sampler_.AddSample(duration_highpri_bytes_through_ * 1000 /
                                      duration_ms);
     limit_bytes_sampler_.AddSample(prev_bytes_per_sec);
   }
+  int64_t new_bytes_per_sec = bytes_sampler_.GetFullValue();
   int32_t ratio = std::max(
       kRatioLower,
       static_cast<int32_t>(
           bytes_sampler_.GetFullValue() * 10 /
           std::max(highpri_bytes_sampler_.GetFullValue(), kHighBytesLower)));
-
-  // in case there are compaction bursts even when online writes are stable
+  // Only adjust threshold when foreground write (flush) flow increases,
+  // because decreasement could also be caused by manual flow control at
+  // application level to alleviate background pressure.
+  new_bytes_per_sec = std::max(
```

Comment:
so when foreground flow decreases, we use long term average throughput as new rate limit? just want to understand the logic.

Reply:
Yes, and this is consistent with before:
When flush flow is steady, i.e. flush_flow.full() == flush_flow.recent(),
old algorithm yields `new_bytes_per_sec = ratio * flush_flow.recent() = compaction_flow.full() / flush_flow.full() * flush_flow.recent() = compaction_flow.full()`